include/mapbox/std/weak.hpp

```diff
@@ -348,6 +348,21 @@ class WeakPtrFactory final {
         };
     }
 
+    /**
+     * @brief Invalidates all existing weak pointers.
+     *
+     * This method is particularly useful, when \c T  class has a user-defined destructor,
+     * that makes the wrapped instance invalid even before entering the \c WeakPtrFactory
+     * destructor.
+     * In this case, clients *MUST* explicitly call \c invalidateWeakPtrs() before freeing any resources.
+     *
+     * Note: After \c invalidateWeakPtrs() is called, \c makeWeakPtr() returns empty weak pointers.
+     */
+    void invalidateWeakPtrs() {
+        if (strong_) strong_->invalidate();
+        strong_.reset();
```

Comment:
This could still fail:

Thread A calls `invalidateWeakPtr` in `WeakPtrFactory`'s destructor. It proceeds until after the `strong_->invalidate()`, but before the `strong_.reset()` call, and gets preempted. At this point, `sharedLocks_` is set to `kInvalidValue` aka `numeric_limits<std::size_t>::max`.

Thread B has a `WeakPtr`. It gets scheduled to run, and calls `.lock()`, which acquires a `std::shared_ptr` of `WeakPtrSharedData`. This is still possible, since the object is still strongly held by the `WeakPtrFactory`. `lock()` then proceeds to call `strong->sharedLock()`. While there's an `assert` in that function that will terminate the program in debug builds, `sharedLock` will happily increment the `sharedLocks_` atomic (which will now be 0 since it overflowed, and unsigned integer overflow is not UB). That means `strong->valid()` will also return `true`, and we'll return a valid `WeakPtrGuard` handle. 💥

Reply:
Thanks for the catch @kkaefer I think the described problem could already happen with the existing code, if the thread B locks a weakPtr on `WeakPtrFactory` destruction :disappointed: 
The updated patch is fixing that, making sure `WeakPtrSharedData::sharedLock()/sharedUnlock()` handles invalid state. PTAL